lib/diff.ml

```diff
@@ -152,32 +174,59 @@ let rec type_item ~typing_env ~name ~reference ~current =
       type_decls ~typing_env ~name ~reference ~current
 
 and type_decls ~typing_env ~name ~reference ~current =
+  let reference, current =
+    if
+      Util.normalized_type_params ~reference:reference.Types.type_params
+        ~current:current.Types.type_params
+    then (reference, current)
+    else Util.normalize_type_decls ~reference ~current
```

Comment:
I like this idea because it means we try to preserve the original names as much as possible. I wonder if we couldn't simplify it into:
```suggestion
    Util.normalize_type_decls ~reference ~current
```
by making `normalize` keep shared names and assign new names where they differ between the two declarations. That way we only process the type parameters once and we can even have more sharing than all or nothing.

What do you think?

Reply:
I am not sure if it would look nice when we print the the textual diff. Also there is a problem that the unique names generated could end up colliding with a name of a type parameter that's the same in the two type declaration. 

For example:
```

type ('a, 'b, ..., 't1) = ...

type ('a, 'c, ..., 't1) = ...
```
Though I think it's possible to resolve the name collision by building the maps first, but still having 't2 before 't1 doesn't look that nice to me.

Reply:
Yeah you're right, it would be too much work to have something correct there and probably not worth the pain!